core/src/iam/verify.rs

```diff
@@ -641,6 +633,51 @@ fn verify_pass(pass: &str, hash: &str) -> Result<(), Error> {
 	}
 }
 
+// Execute the AUTHENTICATE clause for a record access method
+async fn authenticate_record(
+	kvs: &Datastore,
+	session: &Session,
+	authenticate: Value,
+) -> Result<Thing, Error> {
+	match kvs.evaluate(authenticate, session, None).await {
+		Ok(val) => match val.record() {
+			// If the AUTHENTICATE clause returns a record, authentication continues with that record
+			Some(id) => Ok(id),
+			// If the AUTHENTICATE clause returns anything else, authentication fails generically
+			_ => Err(Error::InvalidAuth),
+		},
+		Err(e) => match e {
+			// If the AUTHENTICATE clause throws a specific error, authentication fails with that error
+			Error::Thrown(_) => Err(e),
+			e if *INSECURE_FORWARD_RECORD_ACCESS_ERRORS => Err(e),
```

Comment:
Should we change this to `INSECURE_FORWARD_ACCESS_ERRORS` so we can implement it in other locations too? Just a thought that occurred, can be a separate (future) PR

Perhaps even `INSECURE_FORWARD_AUTH_ERRORS` for all forms of authentication where we cover sensitive errors up with a generic error

Reply:
I think this is a great point. Probably something work thinking about and doing properly on its own PR :+1:

Reply:
Actually, don in [959cc4c](https://github.com/surrealdb/surrealdb/pull/4385/commits/959cc4c34f96bcd81308687778b060727a0261f8).

For now I just opted for the first option you proposed, although we may consider the second in the future :+1: